### ISSUE TITLE
"kind" should be a string and not a number

### ISSUE THREAD

**Issue by alexgleason** [nostr-protocol/nips#511]
Disclaimer: I know this statement is basically insane. This would break literally everything. The goal of this issue is not to convince anyone that we should actually do this, but to create a record of something I believe to be true in the official Nostr repo.

"kind" should be a string instead of a number. Here's why:

1. **There's a limit to how high the number can be.** If we care about JavaScript (which we do because Nostr uses JSON and there are web clients), that value is `9007199254740991`. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER This is **not a problem on its own**, because we will never have that many event types, but it is a problem when combined with the other problems listed below.

2. **It is too easy for numbers to conflict**, especially low numbers. When multiple people are working on a feature that isn't yet an approved NIP, they may choose the same number. One or more features using the same kind could become prevalent even when there is no official NIP yet, causing incompatibilities and grief. Even if this has not happened yet, people _are aware that this could happen_ and feel discouraged from experimenting on Nostr because they know it could happen.

3. **Humans are not good at generating random numbers.** Humans are biased, and have a natural inclination to create events of kind `69`, `420` and `1337`. We would like the number to have a semantic meaning if possible, so we gravatate towards known numbers. Even if we don't, humans in general have been proven to have a hard time coming up with truly random numbers: https://psychology.stackexchange.com/a/3592 To help ease this problem, tools like Nostr Army Knife could generate a random integer within the JavaScript limit for experimental Nostr events, but it would still be obscure and not ideal.

4. **`nostr-protocol/nips` is basically the IANA of Nostr.** This is more of a conclusion drawn from the other points rather than a point on its own. I think centralized development of a decentralized protocol is in general a good thing, but it would be better if there was some formalized way to "do your own thing" within the network. It would make it easier for others to support or ignore your events.

Nostr development is modeled after Bitcoin, but these "limitations" described above are all _desired_ by Bitcoin, because Bitcoin wants to move slowly and not break stuff because money is at stake. Social media is different, and it _wants_ to be moving at a breakneck pace.

**What's the alternative?** Set aside any negative feelings about Bluesky for a moment, because the one thing I think it got right is its naming convention for different kinds of events. It uses reverse-DNS names like `com.example.note`, so that (for example) `io.damus.note` could coexist with `pub.soapbox.note`. This is simpler than ActivityPub's way of using a full URL, and it's not necessarily relient on DNS because it could support nonstandard TLDs. If we really hate that idea and want something more Nosteric instead, the name could be based on a pubkey like `79c2cae114ea28a981e7559b4fe7854a473521a8d22a66bbab9fa248eb820ff6.note`, and possibly even resolvable to something on a relay. I'm sure cryptography people can come up with something simpler and clever.

I don't have the definitive answer about the specific format of the string, I just believe the options I gave are much better than using a number. Instead of trying to poke holes in my alternative solutions, I would rather address the problems above. Or don't - it's not that big of a deal. I just thought it was enough of a problem to write this out after thinking on it for a few months. 🤙

**Comment by staab** [comment]
I basically agree with everything you said, strings would be better, but numbers are fine and what we have.

> it would be better if there was some formalized way to "do your own thing" within the network

This is probably the only part of the post that is actionable. Right now we have https://github.com/nostr-protocol/nips/blob/master/78.md which covers arbitrary key/value storage based on a string identifier. It might be worth adding to that NIP (or somewhere else, CONTRIBUTING.md?) that non-standard features shouldn't use new kinds, but should be prototyped using NIP 78. We would need to expand NIP 78 to include all event types (non-replaceable), but that could reduce the number of conflicts that come up due to conflicting use cases on a single kind. That said, I don't think even that is a big deal, because an attacker can always craft malicious payloads, so clients will always need to code defensively and make no assumptions about the shape of a given event's payload beyond the very basics.

**Comment by mikedilger** [comment]
Eh, I disagree.

There (are, will, or should be) only a small number of kinds.  Because 'kind' is meaningless without interoperability, without buy-in and kind-specific coding by multiple clients/relays.  The community  just isn't going to create a huge number of different code paths.

Say you want to create your own protocol using nostr relays, but you reject all of the nostr kinds, so you create all new ones that you find better, your personal "nostr-next" or "nostr v2".  You can go ahead and do that within the space of 64-bit numbers.  You could even choose randomly and you statistically wouldn't collide with any other 64-bit number ever generated.  That is why they are used for UUIDs and called snowflakes.

The odds of a collision are 1 in 18,446,744,073,709,551,616

**Comment by vitorpamplona** [comment]
Agree with everything. But the missing element is that an integer forces us to interoperate. Otherwise, Nostr would have been just a bunch of "vitor:1", "vitor:2", "damus:3", "iris:4" types. 

Also, be careful about what you wish because this is how you end up with a DID identifier as an event type. Then you will need to code a did resolver and when you finish you will be shipping micro-chains of the historical versioning with-in each event type :) 

PS: We should all take a moment to appreciate how the [JSON spec](https://datatracker.ietf.org/doc/html/rfc8259#section-6) doesn't define either the precision or range for integers and makes statements like "implementations will approximate JSON numbers within the expected precision" without providing any context to what the "expected precision" should be in a given JSON file.

**Comment by alexgleason** [comment]
Closing because this issue was basically just creative writing and I've said my piece. Also, there were some convincing arguments that kind should be a number.

**Comment by alexgleason** [comment]
Actually, this still sucks. Because now Geocaching is mixed with whatever Satlantis is doing.

![Image](https://github.com/user-attachments/assets/445d028a-8ebc-4ec5-b4d2-989bffc4cd5c)

**Comment by arthurfranca** [comment]
@alexgleason or this #1892

**Comment by fiatjaf** [comment]
What if kinds are strings then someone takes the kind "place" and starts publishing GeoJSON blobs, then another app takes the kind "place" and starts publishing human-readable names of places, and another app takes the kind "place" and starts publishing latitudes and longitudes, and another app takes the kind "place" and starts publishing Asciidoc text containing descriptions of places with pictures, and another app takes the kind "place" and starts publishing a bunch of random data (because "place" is actually a verb and they're modifying some state that reads from a different kinds such as "place", "get", "create" etc), and another app takes the kind "place" and starts publishing pictures of palaces (because they had a typo in the kind name but took too long to spot it now they're stuck with it)?

**Comment by mikedilger** [comment]
This is an old issue.  But based on Alex's points 1, 2, 3, and 4, I gathered that he was suggesting that perhaps there should be some way that people could make up their own kind without any significant chance of collision. He mentions things like com.mikedilger.mykind, not things like "place".  But if that were the goal, I would solve it differently with a random 64 bit number.

But I still stand by my first comment from 2023.  I don't think having a central place of agreement on which kind means what is such a bad thing.  It herds the cats.  Being completely distributed here means low interoperability.

**Comment by alexgleason** [comment]
@fiatjaf I don't know. ActivityPub and AtProto use a string kind, and they don't have these problem. Nostr fixes everything they did wrong while not copying what they did right. 😂 Still, I think there are okay solutions. Client-side event validation is the best one I've come up with.

**Comment by fiatjaf** [comment]
What if that's another thing they did wrong?

Anyway, I get your point, there's an argument for using a string, but there are arguments for the number too. Maybe we should also consider using a boolean.

**Comment by alexgleason** [comment]
Update: I fixed it by adding this text into my context file. After this the robot does it correctly:

#### Event Validation

When querying events, if the event kind being returned has required tags or required JSON fields in the content, the events should be filtered through a validator function. This is not generally needed for kinds such as 1, where all tags are optional and the content is freeform text, but is especially useful for custom kinds as well as kinds with strict requirements.

```typescript
// Example validator function for NIP-52 calendar events
function validateCalendarEvent(event: NostrEvent): boolean {
  // Check if it's a calendar event kind
  if (![31922, 31923].includes(event.kind)) return false;

  // Check for required tags according to NIP-52
  const d = event.tags.find(([name]) => name === 'd')?.[1];
  const title = event.tags.find(([name]) => name === 'title')?.[1];
  const start = event.tags.find(([name]) => name === 'start')?.[1];

  // All calendar events require 'd', 'title', and 'start' tags
  if (!d || !title || !start) return false;

  // Additional validation for date-based events (kind 31922)
  if (event.kind === 31922) {
    // start tag should be in YYYY-MM-DD format for date-based events
    const dateRegex = /^\d{4}-\d{2}-\d{2}$/;
    if (!dateRegex.test(start)) return false;
  }

  // Additional validation for time-based events (kind 31923)
  if (event.kind === 31923) {
    // start tag should be a unix timestamp for time-based events
    const timestamp = parseInt(start);
    if (isNaN(timestamp) || timestamp <= 0) return false;
  }

  return true;
}

function useCalendarEvents() {
  const { nostr } = useNostr();

  return useQuery({
    queryKey: ['calendar-events'],
    queryFn: async (c) => {
      const signal = AbortSignal.any([c.signal, AbortSignal.timeout(1500)]);
      const events = await nostr.query([{ kinds: [31922, 31923], limit: 20 }], { signal });
      
      // Filter events through validator to ensure they meet NIP-52 requirements
      return events.filter(validateCalendarEvent);
    },
  });
}
```

**Comment by melvincarvalho** [comment]
> Anyway, I get your point, there's an argument for using a string, but there are arguments for the number too. 

Using **strings** for the `kind` field is a clear win for interoperability, nearly every modern social protocol does the same.

* **Interoperability by default**: let different apps recognise each other’s events without a central allocator.
* **Human-readable & self-documenting**:  A literal name tells you what an event is at a glance; no magic numbers or lookup tables required.
* **Permissionless innovation**: Anyone can ship a new `kind` instantly, experiment in the wild, and iterate without waiting for an official number to be assigned, which has crowded out most developers, during the often multi-year waiting period.

In short, strings make the protocol friendlier for developers, users, and even AI systems, which can infer an event’s purpose directly from its descriptive identifier.

**Comment by melvincarvalho** [comment]
> What if kinds are strings then someone takes the kind "place" and starts publishing GeoJSON blobs, then another app takes the kind "place" and starts publishing human-readable names of places, and another app takes the kind "place" and starts publishing latitudes and longitudes, and another app takes the kind "place"

This problem was solved decades ago with namespacing, the approach ActivityPub and most other modern protocols use.

- Scalability: Plain numbers don’t scale; strings do, and namespaced strings scale indefinitely.

- Decentralisation: Numbers require a central allocator. Namespaced strings let anyone experiment without asking permission.

- DevX: imagine a database where every column must be an integer and you have to petition a registrar, possibly waiting years, for a new “official” number. Most devs would abandon that system for one where they can declare columns by name.

Numbers lock growth behind bureaucracy; namespaced strings unlock it.  A tiny NIP would fix this.

**Comment by melvincarvalho** [comment]
> Agree with everything. But the missing element is that an integer forces us to interoperate. Otherwise, Nostr would have been just a bunch of "vitor:1", "vitor:2", "damus:3", "iris:4" types.

Is not the case, look at how activitypub scales over 100x more than nostr.  Bluesky also 100x nostr with strings.  You need a common agreement on strings or a schema.  These are long solved problems in every day use.  It is the dominant pattern for open protocols, on the social web.  In addition, you get interop out of the box.  If we transitioned to strings, we could:

- alias existing kinds to strings, where necessary:  e.g. 1=Note etc.
- have a small number of centralized kinds that are kept and maintained
- keep the 2xxxx range
- keep the 3xxxx range

Just start with strings which should be good enough for scalilng, and namespaces get layered on top.  Nostr with strings would a very good way to scale the protocol.
